pkg/kube/servicewatcher.go

```diff
@@ -182,6 +192,99 @@ func handleUpdate(oldObj, newObj interface{}, eventCh chan<- event) {
 		}
 	}
 
-	sendEvents(deleted, oldSvc, true)
-	sendEvents(added, newSvc, false)
+	if len(deleted) > 0 {
+		if err := removePortMapping(oldSvc.UID, portTracker); err == nil {
+			log.Debugf("Send remove event for %s", string(oldSvc.UID))
+			sendEvents(deleted, oldSvc, true)
+		}
+	}
+
+	if len(added) > 0 {
+		var err error
+		if existingPortMap := portTracker.Get(string(newSvc.UID)); existingPortMap != nil {
+			err = updatePortMapping(added, existingPortMap, newSvc.UID, portTracker)
+		} else {
+			err = addPortMapping(added, newSvc.UID, portTracker)
+		}
+
+		if err == nil {
+			log.Debugf("Send add event for %s", string(newSvc.UID))
+			sendEvents(added, newSvc, false)
+		}
+	}
+}
+
+func createPortMapping(ports map[int32]corev1.Protocol) (nat.PortMap, error) {
+	portMap := make(nat.PortMap)
+
+	for port, proto := range ports {
+		log.Debugf("create port mapping for port %d, protocol %s", port, proto)
+		portMapKey, err := nat.NewPort(string(proto), strconv.Itoa(int(port)))
+		if err != nil {
+			return nil, err
+		}
+
+		portBinding := nat.PortBinding{
+			HostIP:   net.IPv4zero.String(),
+			HostPort: strconv.Itoa(int(port)),
+		}
+
+		if pb, ok := portMap[portMapKey]; ok {
+			portMap[portMapKey] = append(pb, portBinding)
+		} else {
+			portMap[portMapKey] = []nat.PortBinding{portBinding}
+		}
+	}
+
+	return portMap, nil
+}
+
+func addPortMapping(ports map[int32]corev1.Protocol, serviceUID types.UID, portTracker *tracker.PortTracker) error {
+	portMapAdd, err := createPortMapping(ports)
+	if err != nil {
+		log.Errorf("creating initial port mapping failed: %v", err)
```

Comment:
We log the error and return it, we should do one or another to avoid confusing log outputs. In this case, I suggest we only return and do the logging in the end (the caller of `addPortMapping`).